plugins/Marketplace/PackageList.py

```diff
@@ -91,3 +117,217 @@ def hasFooter(self) -> bool:
+            "packageId": package_id,
+            "handler": self
+        })
+        dialog.show()
+        # place dialog in class such that it does not get remove by garbage collector
+        self._license_dialogs[package_id] = dialog
+
+    @pyqtSlot(str)
+    def onLicenseAccepted(self, package_id: str) -> None:
+        Logger.debug(f"Accepted license for {package_id}")
+        # close dialog
+        dialog = self._license_dialogs.pop(package_id)
+        if dialog is not None:
+            dialog.deleteLater()
+        # install relevant package
+        self._install(package_id)
+
+    @pyqtSlot(str)
+    def onLicenseDeclined(self, package_id: str) -> None:
+        Logger.debug(f"Declined license for {package_id}")
+        # close dialog
+        dialog = self._license_dialogs.pop(package_id)
+        if dialog is not None:
+            dialog.deleteLater()
+        # reset package card
+        package = self.getPackageModel(package_id)
+        package.is_installing = False
+
+    def _requestInstall(self, package_id: str, update: bool = False) -> None:
+        Logger.debug(f"Request installing {package_id}")
+
+        package_path = self._to_install[package_id]
+        license_content = self._manager.getPackageLicense(package_path)
+
+        if not update and license_content is not None:
+            # If installation is not and update, and the packages contains a license then
+            # open dialog, prompting the using to accept the plugin license
+            self._openLicenseDialog(package_id, license_content)
+        else:
+            # Otherwise continue the installation
+            self._install(package_id, update)
+
+    def _install(self, package_id: str, update: bool = False) -> None:
+        Logger.debug(f"Installing {package_id}")
+        package_path = self._to_install.pop(package_id)
+        to_be_installed = self._manager.installPackage(package_path) is not None
+        package = self.getPackageModel(package_id)
+        if package.can_update and to_be_installed:
+            package.can_update = False
+        if update:
+            package.is_updating = False
+        else:
+            package.is_installing = False
+        self.subscribeUserToPackage(package_id, str(package.sdk_version))
+
+    def download(self, package_id: str, url: str, update: bool = False) -> None:
+        """Initiate the download request
+
+        :param package_id: the package identification string
+        :param url: the URL from which the package needs to be obtained
+        :param update: A flag if this is download request is an update process
+        """
+
+        def downloadFinished(reply: "QNetworkReply") -> None:
+            self._downloadFinished(package_id, reply, update)
+
+        def downloadError(reply: "QNetworkReply", error: "QNetworkReply.NetworkError") -> None:
+            self._downloadError(package_id, update, reply, error)
+
+        HttpRequestManager.getInstance().get(
```

Comment:
I think we might need a (second?) `self._ongoing_request = ...` here just to be on the safe side. Though, on the other hand, pressing the install button, immediately switching tabs a few times and then pressing another install button is not likely to happen.

Reply:
I also tried to break it with big packages, such as the CAD plugin and I didn't see any trouble. But sometimes a bit of defensive programming is a good thing. I will change it.